extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/groupby/GroupByQueryKit.java

```diff
@@ -185,13 +186,16 @@ public QueryDefinition makeQueryDefinition(
       );
 
       if (doLimitOrOffset) {
+        final ShuffleSpec finalShuffleSpec = needsFinalShuffling ?
+                                             shuffleSpecFactoryPreAggregation.build(resultClusterBy, true) :
+                                             null;
```

Comment:
I wonder if we can relax this constraint even when we don't have a final shuffle. In intermediate queries, having more partitions will improve the parallelization. Why do we not partition further in case of LIMIT processor?